### PR TITLE
[BugFix] Apply get_cached_tokenizer to the tokenizer setter of LLM

### LINKED ISSUE #5206: [Bug]: Detokenize delay when update vllm from 0.3.0 to 0.4.2
### Your current environment

```text
Collecting environment information...
PyTorch version: 2.3.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (GCC) 9.3.1 20200408 (Red Hat 9.3.1-2)
Clang version: 3.4.2 (tags/RELEASE_34/dot2-final)
CMake version: version 3.29.3
Libc version: glibc-2.17

Python version: 3.10.12 (main, Jul  5 2023, 18:54:27) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-4.18.0-193.28.1.el8_2.x86_64-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA A800-SXM4-80GB
GPU 1: NVIDIA A800-SXM4-80GB

Nvidia driver version: 525.125.06
cuDNN version: Probably one of the following:
/usr/lib64/libcudnn.so.8.0.5
/usr/lib64/libcudnn_adv_infer.so.8.0.5
/usr/lib64/libcudnn_adv_train.so.8.0.5
/usr/lib64/libcudnn_cnn_infer.so.8.0.5
/usr/lib64/libcudnn_cnn_train.so.8.0.5
/usr/lib64/libcudnn_ops_infer.so.8.0.5
/usr/lib64/libcudnn_ops_train.so.8.0.5
/usr/local/cuda/targets/x86_64-linux/lib/libcudnn.so.8.9.5
/usr/local/cuda/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.9.5
/usr/local/cuda/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.9.5
/usr/local/cuda/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.9.5
/usr/local/cuda/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.9.5
/usr/local/cuda/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.9.5
/usr/local/cuda/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.9.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                128
On-line CPU(s) list:   0-127
Thread(s) per core:    2
Core(s) per socket:    32
Socket(s):             2
NUMA node(s):          2
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 106
Model name:            Intel(R) Xeon(R) Platinum 8338C CPU @ 2.60GHz
Stepping:              6
CPU MHz:               3077.183
CPU max MHz:           3500.0000
CPU min MHz:           800.0000
BogoMIPS:              5200.00
Virtualization:        VT-x
L1d cache:             48K
L1i cache:             32K
L2 cache:              1280K
L3 cache:              49152K
NUMA node0 CPU(s):     0-31,64-95
NUMA node1 CPU(s):     32-63,96-127
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid md_clear pconfig flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] nvidia-nccl-cu12==2.20.5
[pip3] torch==2.3.0
[pip3] torchaudio==2.3.0
[pip3] torchvision==0.18.0
[pip3] triton==2.3.0
[pip3] vllm-nccl-cu12==2.18.1.0.4.0
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] nvidia-nccl-cu12          2.20.5                   pypi_0    pypi
[conda] torch                     2.3.0                    pypi_0    pypi
[conda] torchaudio                2.3.0                    pypi_0    pypi
[conda] torchvision               0.18.0                   pypi_0    pypi
[conda] triton                    2.3.0                    pypi_0    pypi
[conda] vllm-nccl-cu12            2.18.1.0.4.0             pypi_0    pypiROCM Version: Could not collect
Neuron SDK Version: N/A
vLLM Version: 0.4.2
vLLM Build Flags:
CUDA Archs: Not Set; ROCm: Disabled; Neuron: Disabled
GPU Topology:
GPU0    GPU1    NIC0    NIC1    NIC2    NIC3    NIC4    NIC5    NIC6    NIC7    NIC8    NIC9    NIC10   CPU Affinity    NUMA Affinity
GPU0     X      NV8     SYS     SYS     SYS     SYS     SYS     SYS     NODE    NODE    PXB     PXB     SYS     32-63,96-127    1
GPU1    NV8      X      SYS     SYS     SYS     SYS     SYS     SYS     NODE    NODE    PXB     PXB     SYS     32-63,96-127    1
NIC0    SYS     SYS      X      PIX     NODE    NODE    NODE    NODE    SYS     SYS     SYS     SYS     NODE
NIC1    SYS     SYS     PIX      X      NODE    NODE    NODE    NODE    SYS     SYS     SYS     SYS     NODE
NIC2    SYS     SYS     NODE    NODE     X      PIX     NODE    NODE    SYS     SYS     SYS     SYS     NODE
NIC3    SYS     SYS     NODE    NODE    PIX      X      NODE    NODE    SYS     SYS     SYS     SYS     NODE
NIC4    SYS     SYS     NODE    NODE    NODE    NODE     X      PIX     SYS     SYS     SYS     SYS     NODE
NIC5    SYS     SYS     NODE    NODE    NODE    NODE    PIX      X      SYS     SYS     SYS     SYS     NODE
NIC6    NODE    NODE    SYS     SYS     SYS     SYS     SYS     SYS      X      PIX     NODE    NODE    SYS
NIC7    NODE    NODE    SYS     SYS     SYS     SYS     SYS     SYS     PIX      X      NODE    NODE    SYS
NIC8    PXB     PXB     SYS     SYS     SYS     SYS     SYS     SYS     NODE    NODE     X      PIX     SYS
NIC9    PXB     PXB     SYS     SYS     SYS     SYS     SYS     SYS     NODE    NODE    PIX      X      SYS
NIC10   SYS     SYS     NODE    NODE    NODE    NODE    NODE    NODE    SYS     SYS     SYS     SYS      X

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_2
  NIC1: mlx5_3
  NIC2: mlx5_4
  NIC3: mlx5_5
  NIC4: mlx5_6
  NIC5: mlx5_7
  NIC6: mlx5_8
  NIC7: mlx5_9
  NIC8: mlx5_10
  NIC9: mlx5_11
  NIC10: mlx5_bond_0
```

### 🐛 Describe the bug

I updated the vllm from 0.3.0 to 0.4.2, both are offical release, while the generation time increased from 45 seconds to 50 minutes on qwen1.5-0.5B model with the AdvertiseGen dataset. 

After compared each phase of generation between the two release version, I found out that the main delay was caused by the function ```detokenize_incrementally in vllm.transformers_utils.detokenizer.py```. 
I added log to record time cost of the upper func, with the result that time cost increased almost 1000 times
```
vllm-0.3.0
INFO 06-03 11:35:10 llm_engine.py:902] detokenize_incrementally cost 1.9311904907226562e-05 seconds
INFO 06-03 11:35:10 llm_engine.py:902] detokenize_incrementally cost 1.9311904907226562e-05 seconds
INFO 06-03 11:35:10 llm_engine.py:902] detokenize_incrementally cost 2.1219253540039062e-05 seconds
INFO 06-03 11:35:10 llm_engine.py:902] detokenize_incrementally cost 2.1219253540039062e-05 seconds
INFO 06-03 11:35:10 llm_engine.py:902] detokenize_incrementally cost 2.09808349609375e-05 seconds
INFO 06-03 11:35:10 llm_engine.py:902] detokenize_incrementally cost 2.1219253540039062e-05 seconds
INFO 06-03 11:35:10 llm_engine.py:902] detokenize_incrementally cost 2.1219253540039062e-05 seconds
INFO 06-03 11:35:10 llm_engine.py:902] detokenize_incrementally cost 4.4345855712890625e-05 seconds
INFO 06-03 11:35:10 llm_engine.py:902] detokenize_incrementally cost 2.1696090698242188e-05 seconds
INFO 06-03 11:35:10 llm_engine.py:902] detokenize_incrementally cost 2.2649765014648438e-05 seconds
```

```
vllm-0.4.2
INFO 06-03 11:49:34 detokenizer.py:314] detokenize_incrementally cost 0.01605677604675293 seconds
INFO 06-03 11:49:34 detokenizer.py:314] detokenize_incrementally cost 0.018158912658691406 seconds
INFO 06-03 11:49:34 detokenizer.py:314] detokenize_incrementally cost 0.019262075424194336 seconds
INFO 06-03 11:49:34 detokenizer.py:314] detokenize_incrementally cost 0.022993803024291992 seconds
INFO 06-03 11:49:34 detokenizer.py:314] detokenize_incrementally cost 0.0168459415435791 seconds
INFO 06-03 11:49:34 detokenizer.py:314] detokenize_incrementally cost 0.014632225036621094 seconds
INFO 06-03 11:49:34 detokenizer.py:314] detokenize_incrementally cost 0.012951850891113281 seconds
INFO 06-03 11:49:34 detokenizer.py:314] detokenize_incrementally cost 0.014124870300292969 seconds
INFO 06-03 11:49:34 detokenizer.py:314] detokenize_incrementally cost 0.018686532974243164 seconds
INFO 06-03 11:49:34 detokenizer.py:314] detokenize_incrementally cost 0.01888108253479004 seconds
INFO 06-03 11:49:34 detokenizer.py:314] detokenize_incrementally cost 0.01627206802368164 seconds
INFO 06-03 11:49:34 detokenizer.py:314] detokenize_incrementally cost 0.018700838088989258 seconds
```

Focus on the fun 'detokenize_incrementally', I find out that ```len(tokenizer)``` is introduced to get the vocab size of the tokenizer, while it's not a nice choice because this func costs much more time than ```tokenizer.vocab_size```. I think we should replace all the ```len(tokenizer)``` with ````tokenizer.vocab_size``` to improve the generation performance

### LINKED ISSUE #5240: [Bug]: Tokenizer setter of LLM without CachedTokenizer adapter
### Your current environment

same to the issue [#5206](https://github.com/vllm-project/vllm/issues/5206)

### 🐛 Describe the bug

As the basic reason of the issue reported by [#5206](https://github.com/vllm-project/vllm/issues/5206), the tokenizer setter of the LLM will override the cached tokenizer inited by llm_engine. 
```python
    def set_tokenizer(
        self,
        tokenizer: Union[PreTrainedTokenizer, PreTrainedTokenizerFast],
    ) -> None:
        self.llm_engine.tokenizer.tokenizer = tokenizer
```
Thus, each time the ```len(tokenizer)``` is called, row ```__len__``` is called rather than the CachedTokenizer's.
To fixed this problem of this issue, the ```get_cached_tokenizer``` adapter should be applied by the tokenizer setter of LLM.

### PR DESCRIPTION
**Pull request by DriverSong**
Change the func to get lengh of tokenizer from ```len(tokenizer)``` to ```tokenizer.vocab_size```

FIX the issue decribed in [#5206](https://github.com/vllm-project/vllm/issues/5206) [#5240](https://github.com/vllm-project/vllm/issues/5240)

FIX #5206
FIX #5240

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by youkaichao** [comment]
Stange, this property should already be cached:

https://github.com/vllm-project/vllm/blob/7a64d24aad69e4d2548aa0bf528d9fe63428ab01/vllm/transformers_utils/tokenizer.py#L17-L55

**Comment by DriverSong** [comment]
> Stange, this property should already be cached:
> 
> https://github.com/vllm-project/vllm/blob/7a64d24aad69e4d2548aa0bf528d9fe63428ab01/vllm/transformers_utils/tokenizer.py#L17-L55

I change the tokenizer by the func ```set_tokenizer``` of llm after LLM init. Maybe we should change this func as below
```python
from vllm.transformers_utils.tokenizer import get_cached_tokenizer

    def set_tokenizer(
        self,
        tokenizer: Union[PreTrainedTokenizer, PreTrainedTokenizerFast],
    ) -> None:
        self.llm_engine.tokenizer.tokenizer = get_cached_tokenizer(tokenizer)
```

**Comment by DriverSong** [comment]
@youkaichao

**Comment by DriverSong** [comment]
@youkaichao 
I've tested upper change of ```set_tokenizer ```  on qwen1.5-0.5B with the AdvertiseGen dataset., the generate time decreased to 35 seconds. 
```shell
master-0: INFO 06-03 18:38:08 model_runner.py:175] Loading model weights took 0.9018 GB
master-0: INFO 06-03 18:38:08 gpu_executor.py:114] # GPU blocks: 46845, # CPU blocks: 2730
master-0: INFO 06-03 18:38:11 model_runner.py:937] Capturing the model for CUDA graphs. This may lead to unexpected consequences if the model is not static. To run the model in eager mode, set 'enforce_eager=True' or use '--enforce-eager' in the CLI.
master-0: INFO 06-03 18:38:11 model_runner.py:941] CUDA graphs can take additional 1~3 GiB memory per GPU. If you are running out of memory, consider decreasing `gpu_memory_utilization` or enforcing eager mode. You can also reduce the `max_num_seqs` as needed to decrease memory usage.
master-0: INFO 06-03 18:38:18 model_runner.py:1017] Graph capturing finished in 7 secs.
Processed prompts: 100%|██████████| 1070/1070 [00:35<00:00, 29.73it/s]
master-0: [2024-06-03 18:38:57,974] [INFO] [launch.py:351:main] Process 2497 exits successfully.
```

File: vllm/transformers_utils/tokenizer.py [hunk 1]
```diff
@@ -29,7 +29,7 @@ def get_cached_tokenizer(
     tokenizer_all_special_tokens_extended = (
         tokenizer.all_special_tokens_extended)
     tokenizer_all_special_tokens = set(tokenizer.all_special_tokens)
-    tokenizer_len = len(tokenizer)
```
Comment (youkaichao) [inline vllm/transformers_utils/tokenizer.py:32]: why is this change necessary? `len(tokenizer)` might also contain special tokens, while `tokenizer.vocab_size` might not.

File: vllm/entrypoints/llm.py [hunk 2]
```diff
@@ -152,7 +153,7 @@ def set_tokenizer(
         self,
         tokenizer: Union[PreTrainedTokenizer, PreTrainedTokenizerFast],
     ) -> None:
-        self.llm_engine.tokenizer.tokenizer = tokenizer
```
Comment (youkaichao) [inline vllm/entrypoints/llm.py:155]: better add a check, if the tokenizer is already cached, then just set it.

**Comment by youkaichao** [comment]
LGTM in general, please address the comments.

File: vllm/transformers_utils/tokenizer.py (hunk 1 above)
Comment (DriverSong) [reply vllm/transformers_utils/tokenizer.py]: this change has been reverted

File: vllm/entrypoints/llm.py (hunk 2 above)
Comment (DriverSong) [reply vllm/entrypoints/llm.py:155]: A check is added. While the CachedTokenizer is dynamic,  I have no choice but compare the class name. Is there any good ideal?

**Review by mgoin — APPROVED** [review]
